src/ctap2/mod.rs

```diff
@@ -809,5 +813,497 @@ pub(crate) fn set_or_change_pin_helper(
+    let mut current_rp = 0;
+    let mut skip_puap = false;
+    let mut pin_uv_auth_result = PinUvAuthResult::NoAuthRequired;
+    let mut pin = None;
+    while alive() {
+        if !skip_puap {
+            pin_uv_auth_result = match determine_puap_if_needed(
+                &mut cred_management,
+                dev,
+                skip_uv,
+                PinUvAuthTokenPermission::CredentialManagement,
+                UserVerificationRequirement::Preferred,
+                &status,
+                &callback,
+                alive,
+                &mut pin,
+            ) {
+                Ok(r) => r,
+                Err(()) => {
+                    return false;
+                }
+            };
+        }
+
+        debug!("------------------------------------------------------------------");
+        debug!("{cred_management:?} using {pin_uv_auth_result:?}");
+        debug!("------------------------------------------------------------------");
+
+        let resp = dev.send_cbor_cancellable(&cred_management, alive);
+        match resp {
+            Ok(result) => {
+                skip_puap = true;
+                match cred_management.subcommand {
+                    CredManagementCommand::GetCredsMetadata => {
+                        let existing_resident_credentials_count =
+                            unwrap_option!(result.existing_resident_credentials_count, callback);
+                        let max_possible_remaining_resident_credentials_count = unwrap_option!(
+                            result.max_possible_remaining_resident_credentials_count,
+                            callback
+                        );
+                        credential_result.existing_resident_credentials_count =
+                            existing_resident_credentials_count;
+                        credential_result.max_possible_remaining_resident_credentials_count =
+                            max_possible_remaining_resident_credentials_count;
+                        if existing_resident_credentials_count > 0 {
+                            cred_management.subcommand = CredManagementCommand::EnumerateRPsBegin;
+                            // We have to regenerate PUAP here. PUAT hasn't changed, but the content
+                            // of the command has changed, and that is part of the PUAP-calculation
+                            unwrap_result!(
+                                cred_management.set_pin_uv_auth_param(
+                                    pin_uv_auth_result.get_pin_uv_auth_token()
+                                ),
+                                callback
+                            );
+                            continue;
+                        } else {
+                            // This token doesn't have any resident keys, but its not an error,
+                            // so we return an Ok with an empty list.
+                            callback.call(Ok(ManageResult::CredManagement(
```

Comment:
All of these `callback.call(Ok(...))` lines should be status updates. After each one, we should continue from the top of the loop where we send a `Sender<CredManagementCommand>`.

In this particular case, where we're sending an empty credential list back, the user won't have much to do. They can just drop the `Sender` to signal they're done, and then we can `callback.call(Ok(()))`.